src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/AbstractQueryAndScanMapper.java

```diff
@@ -0,0 +1,292 @@
+import static com.salesforce.dynamodbv2.mt.mappers.sharedtable.impl.MappingUtils.getNextValuePlaceholder;
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import com.amazonaws.services.dynamodbv2.model.ComparisonOperator;
+import com.amazonaws.services.dynamodbv2.model.Condition;
+import com.amazonaws.services.dynamodbv2.model.QueryRequest;
+import com.amazonaws.services.dynamodbv2.model.ScanRequest;
+import com.amazonaws.services.dynamodbv2.model.ScanResult;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.salesforce.dynamodbv2.mt.mappers.metadata.DynamoTableDescription;
+import com.salesforce.dynamodbv2.mt.mappers.sharedtable.impl.RequestWrapper.AbstractRequestWrapper;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+abstract class AbstractQueryAndScanMapper implements QueryAndScanMapper {
+
+    protected final DynamoTableDescription virtualTable;
+    private final Function<String, RequestIndex> indexLookup;
+    private final ConditionMapper conditionMapper;
+    private final ItemMapper itemMapper;
+
+    AbstractQueryAndScanMapper(DynamoTableDescription virtualTable,
+                               Function<String, RequestIndex> indexLookup,
+                               ConditionMapper conditionMapper,
+                               ItemMapper itemMapper) {
+        this.virtualTable = virtualTable;
+        this.indexLookup = indexLookup;
+        this.conditionMapper = conditionMapper;
+        this.itemMapper = itemMapper;
+    }
+
+    @Override
+    public void apply(QueryRequest queryRequest) {
+        validateQueryRequest(queryRequest);
+        convertLegacyExpression(new QueryLegacyConditionRequestWrapper(queryRequest), queryRequest.getKeyConditions());
+
+        // map index name if needed
+        RequestIndex requestIndex = getRequestIndexAndMapIndexName(new QueryRequestWrapper(queryRequest, null, null));
+
+        // map key condition
+        QueryRequestWrapper keyConditionExpressionRequest = new QueryRequestWrapper(queryRequest,
+            queryRequest::getKeyConditionExpression, queryRequest::setKeyConditionExpression);
+        conditionMapper.applyToKeyCondition(keyConditionExpressionRequest, requestIndex);
+
+        // map filter expression if there is one
+        QueryRequestWrapper filterExpressionRequest = new QueryRequestWrapper(queryRequest,
+            queryRequest::getFilterExpression, queryRequest::setFilterExpression);
+        applyToFilterExpression(filterExpressionRequest);
+
+        // map exclusive start key if there is one
+        applyExclusiveStartKey(filterExpressionRequest, requestIndex);
+    }
+
```

Comment:
The interface now has executeScan() instead of an apply() that takes a ScanRequest, since for hash partitioning does things differently and runs physical QueryRequests instead.